### ISSUE TITLE
[REQUEST]Support for multiple node inference?

### ISSUE THREAD

**Issue by sleepwalker2017** [microsoft/DeepSpeed#4704]
Hi, I want to run one LLM model using multiple machines.

On one node, I want to use tensor parallel to speedup.

Within multiple nodes, I want to use pipeline parallel.

Is this supported? If so, is there any documentation? I didn't find it.

If not, any documentation for inference using multiple nodes?

**Comment by tjruwase** [comment]
@sleepwalker2017, yes DeepSpeed supports combinations of tensor and pipeline parallelism, a.k.a., [3D parallelism](https://huggingface.co/blog/bloom-megatron-deepspeed#zero-dppptp). However, these techniques are not natively provided by DeepSpeed, instead you need to implement them in your model. You can explore example implementations from [deepspeed](https://github.com/microsoft/Megatron-DeepSpeed) and [bigscience](https://github.com/bigscience-workshop/Megatron-DeepSpeed).

**Comment by sleepwalker2017** [comment]
> @sleepwalker2017, yes DeepSpeed supports combinations of tensor and pipeline parallelism, a.k.a., [3D parallelism](https://huggingface.co/blog/bloom-megatron-deepspeed#zero-dppptp). However, these techniques are not natively provided by DeepSpeed, instead you need to implement them in your model. You can explore example implementations from [deepspeed](https://github.com/microsoft/Megatron-DeepSpeed) and [bigscience](https://github.com/bigscience-workshop/Megatron-DeepSpeed).

Thank you, seems this needs lots of modifications to the model.  Does that mean we need to modify the model code to add some communication primitives? 

Is the tensor parallelism between nodes natively supported in deepspeed? Any example for this feature?

**Comment by tjruwase** [comment]
Sorry, I just noticed you are asking for inference, not training. For inference, DeepSpeed provides native support for tensor parallelism. Please see the following: 
 
1. https://github.com/microsoft/DeepSpeed/tree/master/blogs/deepspeed-fastgen
2. https://www.deepspeed.ai/2022/10/10/mii.html
3. https://www.deepspeed.ai/2021/03/15/inference-kernel-optimization.html

**Comment by tjruwase** [comment]
> Is the tensor parallelism between nodes natively supported in deepspeed? Any example for this feature?

Yes, this should natively work for inference, but performance will be poor because of cross-node communication.

**Comment by sleepwalker2017** [comment]
> > Is the tensor parallelism between nodes natively supported in deepspeed? Any example for this feature?
> 
> Yes, this should natively work for inference, but performance will be poor because of cross-node communication.

For inference, the pipeline parallelism is not supported ?  Is that the fact?

And also, I have tried out deepspeed for local inference on a single machine. The usage is simple.

But I didn't find some documentation for deploying model across machines. Any docs for that?

Thank you.

**Comment by tjruwase** [comment]
Please ensure your environment is setup for [multi-node execution](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) and try the same steps for single machine inference.

**Comment by sleepwalker2017** [comment]
> Please ensure your environment is setup for [multi-node execution](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) and try the same steps for single machine inference.

That's exactly what I need, Thank you!

**Comment by sleepwalker2017** [comment]
> Please ensure your environment is setup for [multi-node execution](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) and try the same steps for single machine inference.

Hi, does deepspeed support PP for inference?

**Comment by Sri-Vadlamani** [comment]
Hi @sleepwalker2017 was wondering if you were able to accomplish model inference using deepspeed with multinode. I am working on similar problem but couldn't find any proper documentation. Thank you
